onnx/onnx.in.proto

```diff
@@ -379,10 +379,39 @@ message TensorShapeProto {
       int64 dim_value = 1;
       string dim_param = 2;   // namespace Shape
     };
+    // Standard denotation can optionally be used to denote tensor
+    // dimensions with standard semantic descriptions to ensure
+    // that operations are applied to the correct axis of a tensor.
+    optional string denotation = 3;
   };
   repeated Dimension dim = 1;
 }
 
+// A set of pre-defined constants to be used as values for
+// the standard denotation field in TensorShapeProto.Dimension
+// for semantic description of the tensor dimension.
+message DenotationConstProto {
```

Comment:
I don't quite understand this ... is this meant to represent an "enum"? This is defined as a message, but is not used anywhere.

Reply:
Yes. (cf @dzhulgakov 's comment on Jan 10 [here](https://github.com/onnx/onnx/issues/406) and his comment in this thread regarding using string constants over enum).

Reply:
I understand the intention ... but was wondering about the Protobuf mechanism used for this purpose here. I see that default values are not supported in Proto3, which could be a problem.

Reply:
I think dzhulgakov said that ONNX is locked to protobuf 2.

Reply:
@gramalingam @tjingrant Yes we are locked to proto 2 syntax (note the syntax version is different from the protobuf library version).

Reply:
@bddppq This is causing a regression in [MLdotNet](https://github.com/dotnet/machinelearning/) C# based machine learning toolkit which was relying on proto3 format for C# code gen from proto files. I'm now getting "onnx-ml.proto3: Explicit default values are not allowed in proto3." Is there a work around that I can use to unblock myself?  If ONNX is locked on proto2 format then can you please clarify why proto3 files are being produced?

Reply:
The agreement we reached last year was to use the common subset of proto 2 and 3 semantics. Any protobuf feature that cannot be precisely represented in both v2 and v3 should not be used.